### PR TITLE
feat(sns): Port Swap's periodic tasks from heartbeats to timers

### PR DESCRIPTION
**Pull request by aterga**
This PR ports Swap's periodic tasks from heartbeats to timers, making use of the fact that Swap has recently been ported to IC-CDK. For Swap's periodic tasks, timers are better than heartbeats for two reasons:

1. Swap can run them less frequently (I propose once a minute vs. once every round), which should start saving resources right away.
2. Swap can cancel the scheduling of any periodic tasks after reaching the terminal lifecycle state, which should drastically save subnet resources long-term.

### CODE REVIEW

File: rs/sns/swap/canister/canister.rs [hunk 1]
```diff
@@ -277,6 +276,29 @@ fn create_real_icp_ledger(id: CanisterId) -> IcpLedgerCanister<CdkRuntime> {
...
 
+fn stop_scheduling() {
+    TIMER_ID.with(|saved_timer_id| {
+        let timer_id = saved_timer_id.borrow();
+        ic_cdk_timers::clear_timer(*timer_id);
+    });
+}
+
+async fn run_periodic_tasks() {
+    swap_mut().run_periodic_tasks(now_fn, stop_scheduling).await;
```
Comment (max-dfinity) [inline rs/sns/swap/canister/canister.rs:287]:
```suggestion
   if !swap().requires_periodic_tasks() {
      stop_scheduling();
   }
    swap_mut().run_periodic_tasks(now_fn).await;
```

**Review by max-dfinity — APPROVED** [review]

**Comment by andrew-lee-work** [comment]
LGTM. I left a comment asking about the spawn in the timer task. Also we discussed elsewhere about having a manual reset functionality in case the timer is broken. Also the additional instruction cost of the timer_executor from making an inter canister call should not be significant enough to cause a problem, given that all tasks until the first inter-canister call in auto_finalize are small, and auto-finalize can be called manually if it fails in the timer.

File: rs/sns/swap/canister/canister.rs [hunk 2]
```diff
@@ -277,6 +276,34 @@ fn create_real_icp_ledger(id: CanisterId) -> IcpLedgerCanister<CdkRuntime> {
...
+            let timer_id = saved_timer_id.borrow();
+            ic_cdk_timers::clear_timer(*timer_id);
+        });
+    }
+}
+
+fn init_timers() {
+    if swap().requires_periodic_tasks() {
+        let timer_id = ic_cdk_timers::set_timer_interval(RUN_PERIODIC_TASKS_INTERVAL, || {
+            ic_cdk::spawn(run_periodic_tasks())
```
[the remark this answers is not in the document]
Comment (aterga) [reply rs/sns/swap/canister/canister.rs:298]: @jasonz-dfinity, could you maybe help answer this question, please?

[the remark this answers is not in the document]
Comment (max-dfinity) [reply rs/sns/swap/canister/canister.rs:298]: I think it's just because run_periodic_tasks is async.  Is there a better way to do that?

**Pull request merged**